### PR TITLE
K8SPXC-167 Add users secrets controller

### PR DESCRIPTION
**Pull request by nonemax**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| self-healing | passed |
| init-deploy | failed |
| big-data | passed |
| recreate | passed |
| scaling | passed |
| storage | passed |
| scaling-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| demand-backup | passed |
| upgrade | passed |
| operator-self-healing | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| one-pod | passed |
| affinity | passed |
| scheduled-backup | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |

File: pkg/pxc/users/users.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+								Name:  "PXC_CONNS",
+								Value: conns,
+							},
+							{
+								Name:  "PXC-ROOT-PASS",
+								Value: rootPass,
+							},
+						},
+					},
+				},
```
Comment (delgod) [inline pkg/pxc/users/users.go]: ContainerSecurityContext, PodSecurityContext, Tolerations, SchedulerName, PriorityClassName - should be inherided from PXC.
also resources limits/requests should be via separate section

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| self-healing | failed |
| scaling | passed |
| init-deploy | passed |
| big-data | passed |
| scaling-proxysql | passed |
| storage | passed |
| upgrade | passed |
| demand-backup | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| scheduled-backup | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| init-deploy | failed |
| big-data | failed |
| scaling | passed |
| self-healing | failed |
| self-healing-advanced | failed |
| storage | failed |
| operator-self-healing | failed |
| scaling-proxysql | passed |
| demand-backup | passed |
| upgrade | passed |
| one-pod | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring | passed |
| auto-tuning | passed |
| scheduled-backup | failed |
| monitoring-2-0 | passed |
| affinity | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| self-healing | passed |
| big-data | passed |
| init-deploy | failed |
| scaling | passed |
| storage | passed |
| scaling-proxysql | passed |
| demand-backup | passed |
| self-healing-advanced | failed |
| limits | passed |
| operator-self-healing | passed |
| monitoring | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| auto-tuning | passed |
| affinity | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |

File: pkg/pxc/users/users.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+var log = logf.Log.WithName("users-manager")
+
+func Job(cr *api.PerconaXtraDBCluster) *batchv1.Job {
+	labels := make(map[string]string)
+	for key, value := range cr.Spec.Users.Labels {
+		labels[key] = value
+	}
+	labels["type"] = "user-manager"
+	labels["cluster"] = cr.Name
+	labels["job-name"] = cr.Name + "-pxc-user-manager"
```
Comment (delgod) [inline pkg/pxc/users/users.go]: what if we need to apply different secrets simultaneously?
will it work fine with the same job name?

File: pkg/pxc/users/users.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+						Command:   []string{"user-manager"},
+						Resources: resources,
+					},
+				},
+				Volumes: []corev1.Volume{
+					{
+						Name: "userssecret",
+						VolumeSource: corev1.VolumeSource{
+							Secret: &corev1.SecretVolumeSource{
+								SecretName: "secret-for-users",
```
Comment (delgod) [inline pkg/pxc/users/users.go]: I believe name secret should be passed from CR, not hardcoded

File: pkg/pxc/users/users.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+var log = logf.Log.WithName("users-manager")
+
+func Job(cr *api.PerconaXtraDBCluster) *batchv1.Job {
+	labels := make(map[string]string)
+	for key, value := range cr.Spec.Users.Labels {
+		labels[key] = value
+	}
+	labels["type"] = "user-manager"
+	labels["cluster"] = cr.Name
```
Comment (delgod) [inline pkg/pxc/users/users.go]:
```suggestion
	labels["app.kubernetes.io/name"] = "percona-xtradb-cluster"
	labels["app.kubernetes.io/instance"] = cr.Name
	labels["app.kubernetes.io/component"] = "user-manager"
	labels["app.kubernetes.io/managed-by"] = "percona-xtradb-cluster-operator"
	labels["app.kubernetes.io/part-of"] = "percona-xtradb-cluster"
```

File: pkg/pxc/users/users.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+							{
+								MountPath: "/data",
+								Name:      "userssecret",
+								ReadOnly:  true,
+							},
+						},
+						Env: []corev1.EnvVar{
+							{
+								Name:  "PXC_SERVICE",
+								Value: conns,
```
Comment (delgod) [inline pkg/pxc/users/users.go]:
```suggestion
								Value: cr.Name+"-pxc",
```

File: pkg/controller/pxc/users.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+			if usersSecretObj.Annotations["status"] == "succeded" || usersSecretObj.Annotations["status"] == "failed" {
+				return nil
+			}
+		}
+	}
+
+	if len(usersSecretObj.Annotations) == 0 {
+		usersSecretObj.Annotations = make(map[string]string)
+	}
+	usersSecretObj.Annotations["last-applied"] = newHash
```
Comment (delgod) [inline pkg/controller/pxc/users.go]: avoid a situation when status was `succeded` for old hash
```suggestion
	usersSecretObj.Annotations["status"] = "applying"
	usersSecretObj.Annotations["last-applied"] = newHash
```
lines 113-119 can be removed

File: pkg/controller/pxc/users.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+			return errors.Wrapf(err, "create job '%s'", job.Name)
+		}
+		if len(lastAppliedHash) > 0 {
+			usersSecretObj.Annotations["status"] = "applying"
+			err = r.client.Update(context.TODO(), &usersSecretObj)
+			if err != nil {
+				return errors.Wrap(err, "update secret status")
+			}
+		}
+		time.Sleep(time.Second * 5)
```
Comment (delgod) [inline pkg/controller/pxc/users.go]: why it is needed?

File: pkg/controller/pxc/users.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+		}
+		if currentJob.Status.Failed > 0 {
+			status = "failed"
+		}
+		usersSecretObj.Annotations["status"] = status
+		err = r.client.Update(context.TODO(), &usersSecretObj)
+		if err != nil {
+			return errors.Wrap(err, "update secret status")
+		}
+		err = r.client.Delete(context.TODO(), currentJob)
```
Comment (delgod) [inline pkg/controller/pxc/users.go]: how it is possible to debug if a failed job deleted

File: deploy/cr.yaml [hunk 9]
```diff
@@ -8,6 +8,8 @@ metadata:
 #    - delete-pxc-pvc
 spec:
   secretsName: my-cluster-secrets
+#  usersSecrets:
+#  - "secret-for-users"
```
Comment (delgod) [inline deploy/cr.yaml]: we need to add example into `deploy/app-users.yaml`

File: pkg/pxc/users/users.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+							},
+							{
+								Name:  "MYSQL_ROOT_PASSWORD",
+								Value: rootPass,
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/pxc/users/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: pkg/controller/pxc/users.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	if err != nil && k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrapf(err, "get cluster secret '%s'", cr.Spec.SecretsName)
+	}
+	operator := corev1.Pod{}
+	err = r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      os.Getenv("HOSTNAME"),
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/controller/pxc/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: pkg/pxc/users/users.go (hunk 2 above)
Comment (nonemax) [reply pkg/pxc/users/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: pkg/pxc/users/users.go (hunk 3 above)
Comment (nonemax) [reply pkg/pxc/users/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: pkg/pxc/users/users.go (hunk 4 above)
Comment (nonemax) [reply pkg/pxc/users/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: pkg/pxc/users/users.go (hunk 5 above)
Comment (nonemax) [reply pkg/pxc/users/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: pkg/controller/pxc/users.go (hunk 6 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: pkg/controller/pxc/users.go (hunk 7 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: removed https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: pkg/controller/pxc/users.go (hunk 8 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b

File: deploy/cr.yaml (hunk 9 above)
Comment (nonemax) [reply deploy/cr.yaml]: added https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/a457d270418b7ddce6f9b7ec3f859fda1bd21b36

File: pkg/pxc/users/users.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+				Labels:      job.Labels,
+				Annotations: cr.Spec.Users.Annotations,
+			},
+			Spec: corev1.PodSpec{
+				RestartPolicy:    corev1.RestartPolicyNever,
+				ImagePullSecrets: imagePullSecrets,
+				Containers: []corev1.Container{
+					{
+						Name:            job.Name,
+						Image:           image + "-docker",
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/pxc/users/users.go]: removed https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/cc24aed0d4db5ef90dfc27e3449ee603322126de

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| self-healing | passed |
| big-data | passed |
| init-deploy | passed |
| recreate | passed |
| scaling | passed |
| storage | failed |
| scaling-proxysql | passed |
| upgrade | passed |
| self-healing-advanced | failed |
| demand-backup | failed |
| operator-self-healing | passed |
| limits | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| scheduled-backup | passed |
| auto-tuning | passed |
| affinity | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| self-healing | passed |
| scaling | passed |
| init-deploy | failed |
| big-data | failed |
| recreate | passed |
| storage | passed |
| scaling-proxysql | passed |
| limits | passed |
| self-healing-advanced | failed |
| demand-backup | failed |
| monitoring | passed |
| upgrade | failed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| affinity | passed |
| scheduled-backup | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| big-data | passed |
| scaling | passed |
| recreate | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| storage | passed |
| self-healing-advanced | failed |
| limits | passed |
| demand-backup | passed |
| upgrade | passed |
| operator-self-healing | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| one-pod | failed |
| monitoring-2-0 | passed |
| scheduled-backup | failed |
| auto-tuning | passed |
| affinity | passed |
| proxysql-sidecar-res-limits | passed |

File: pkg/controller/pxc/users.go [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+func (r *ReconcilePerconaXtraDBCluster) reconcileUsers(cr *api.PerconaXtraDBCluster) error {
+	if cr.Status.Status != api.AppStateReady {
+		return nil
+	}
+
+	if len(cr.Spec.Users.Secrets) == 0 {
+		return nil
+	}
+	secretObj := corev1.Secret{}
```
Comment (delgod) [inline pkg/controller/pxc/users.go]: why it is needed? looks like don't use it at all

File: pkg/controller/pxc/users.go [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+		&secretObj,
+	)
+	if err != nil && k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrapf(err, "get cluster secret '%s'", cr.Spec.SecretsName)
+	}
+
+	for _, secretName := range cr.Spec.Users.Secrets {
+		operator := corev1.Pod{}
```
Comment (delgod) [inline pkg/controller/pxc/users.go]: can we fetch it only if when we run user-manager (not on every reconcile loop)

File: pkg/controller/pxc/users.go [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	job := users.Job(cr, secretName)
+	job.Spec = users.JobSpec(secretName, containerImage, job, cr, imagePullSecrets)
+
+	currentJob := new(batchv1.Job)
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: job.Name, Namespace: cr.Namespace}, currentJob)
+	if err != nil && k8serrors.IsNotFound(err) {
+		err = r.client.Create(context.TODO(), job)
```
Comment (delgod) [inline pkg/controller/pxc/users.go]: looks like we create job only once, if secret changed second time it is not gona created

Comment (nonemax) [reply pkg/controller/pxc/users.go]: now we removing "out of date" job https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/74d8e58743302032ea789cd46dfb590f31384fef

File: pkg/controller/pxc/users.go (hunk 14 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/74d8e58743302032ea789cd46dfb590f31384fef

File: pkg/controller/pxc/users.go (hunk 13 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: removed https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/74d8e58743302032ea789cd46dfb590f31384fef

File: pkg/pxc/users/users.go [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+	for key, value := range cr.Spec.Users.Labels {
+		labels[key] = value
+	}
+
+	labels["app.kubernetes.io/name"] = "percona-xtradb-cluster"
+	labels["app.kubernetes.io/instance"] = cr.Name
+	labels["app.kubernetes.io/component"] = "user-manager"
+	labels["app.kubernetes.io/managed-by"] = "percona-xtradb-cluster-operator"
+	labels["app.kubernetes.io/part-of"] = "percona-xtradb-cluster"
+	labels["job-name"] = secretName + "-" + cr.Name + "-pxc-user-manager"
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/pxc/users/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/74d8e58743302032ea789cd46dfb590f31384fef

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling | passed |
| big-data | passed |
| scaling-proxysql | passed |
| storage | failed |
| self-healing-advanced | passed |
| upgrade | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| one-pod | passed |
| limits | passed |
| monitoring | passed |
| auto-tuning | passed |
| scheduled-backup | failed |
| monitoring-2-0 | passed |
| affinity | passed |
| proxysql-sidecar-res-limits | passed |

File: pkg/controller/pxc/users.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+		},
+		&operatorPod,
+	)
+	if err != nil {
+		return errors.Wrap(err, "get operator deployment")
+	}
+	containerImage := operatorPod.Spec.Containers[0].Image
+	imagePullSecrets := operatorPod.Spec.ImagePullSecrets
+	job := users.Job(cr, secretName, newHash)
+	job.Spec = users.JobSpec(secretName, containerImage, job, cr, imagePullSecrets)
```
Comment (delgod) [inline pkg/controller/pxc/users.go]: I believe this code is needed only if we Create new job (between lines 94-95).
Job name can be calculated outside of `Job` function

File: pkg/pxc/users/users.go [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/pxc/users/users.go]: job name generation added https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/1614b84316670d6c6c63719f51d8ccbea626a485

File: pkg/controller/pxc/users.go (hunk 17 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/1614b84316670d6c6c63719f51d8ccbea626a485

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | failed |
| recreate | failed |
| scaling | failed |
| self-healing | failed |
| init-deploy | failed |
| demand-backup | failed |
| storage | failed |
| scaling-proxysql | failed |
| self-healing-advanced | failed |
| limits | passed |
| operator-self-healing | failed |
| upgrade | passed |
| scheduled-backup | failed |
| monitoring | failed |
| one-pod | passed |
| monitoring-2-0 | failed |
| auto-tuning | failed |
| affinity | passed |
| proxysql-sidecar-res-limits | failed |
| upgrade-consistency | passed |

File: pkg/controller/pxc/users.go (hunk 17 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/1614b84316670d6c6c63719f51d8ccbea626a485

File: pkg/pxc/users/users.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/pxc/users/users.go]: name generator added https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/1614b84316670d6c6c63719f51d8ccbea626a485

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| recreate | passed |
| scaling | passed |
| init-deploy | passed |
| self-healing | failed |
| self-healing-advanced | failed |
| operator-self-healing | passed |
| scaling-proxysql | passed |
| storage | failed |
| demand-backup | passed |
| upgrade | failed |
| one-pod | passed |
| upgrade-consistency | passed |
| limits | passed |
| monitoring | passed |
| auto-tuning | passed |
| monitoring-2-0 | passed |
| scheduled-backup | passed |
| affinity | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| scaling | passed |
| self-healing | failed |
| init-deploy | failed |
| recreate | passed |
| self-healing-advanced | passed |
| scaling-proxysql | passed |
| storage | passed |
| upgrade | failed |
| limits | passed |
| operator-self-healing | passed |
| demand-backup | passed |
| upgrade-consistency | passed |
| monitoring | passed |
| one-pod | passed |
| auto-tuning | passed |
| monitoring-2-0 | passed |
| scheduled-backup | passed |
| affinity | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by delgod** [comment]
discussed offline, should be reworked

**Pull request closed without merging**